ENIGMAsystem/SHELL/Audio_Systems/SDL/SDLadvanced.cpp

```diff
@@ -0,0 +1,146 @@
+    SoundChannel* sc = new SoundChannel();
+    sc->mchunk = snd.mc;
+    sc->priority = priority;
+    sc->soundIndex = index;
+    sound_channels.push_back(sc);
+    if (Mix_PlayChannel(sound_channels.size() -1,snd.mc, loop)==-1) {
+      DEBUG_MESSAGE( std::string("Mix_PlayChannel : ")+std::string(Mix_GetError()), MESSAGE_TYPE::M_ERROR);
+    }
+    return (sound_channels.size() -1 + AUDIO_CHANNEL_OFFSET);
+  }
+  else {
+    for(size_t i =0; i <sound_channels.size();i++) {
+      if (Mix_Paused(i) == 1 || sound_channels[i]->priority < priority) {
+        if (Mix_PlayChannel(i,snd.mc, loop)==-1) {
+          DEBUG_MESSAGE( std::string("Mix_PlayChannel : ")+std::string(Mix_GetError()), MESSAGE_TYPE::M_ERROR);
+        }
+        sound_channels[i]->mchunk = snd.mc;
+        sound_channels[i]->priority = priority;
+        sound_channels[i]->soundIndex = index;
+        return (i + AUDIO_CHANNEL_OFFSET);
+      }
+    }
+  }
+  return -1;
+}
+  
+bool audio_is_playing(int index) {
+  if (index >= AUDIO_CHANNEL_OFFSET) {
+    const Sound& snd = sounds.get(index - AUDIO_CHANNEL_OFFSET);
+    if (sound_channels[index - AUDIO_CHANNEL_OFFSET]->mchunk == snd.mc && Mix_Playing(index - AUDIO_CHANNEL_OFFSET) == 1) { return true; }
+  }
+  const Sound& snd = sounds.get(index);
+  for(size_t i =0; i <sound_channels.size();i++) {
+    if (sound_channels[i]->mchunk == snd.mc && Mix_Playing(i) == 1) { return true; }
+  }
+  return false;
+}
+
+bool audio_is_paused(int index) {
+  if (index >= AUDIO_CHANNEL_OFFSET) {
+    const Sound& snd = sounds.get(index - AUDIO_CHANNEL_OFFSET);
+    if (sound_channels[index - AUDIO_CHANNEL_OFFSET]->mchunk == snd.mc && Mix_Paused(index - AUDIO_CHANNEL_OFFSET) == 1) { return true; }
+  }
+  const Sound& snd = sounds.get(index);
+  for(size_t i =0; i <sound_channels.size();i++) {
+    if (sound_channels[i]->mchunk == snd.mc && Mix_Paused(i) == 1) { return true; }
+  }
+  return false;
+}
+
+void audio_stop_sound(int index) {
+  if (index >= AUDIO_CHANNEL_OFFSET) {
+    const Sound& snd = sounds.get(index - AUDIO_CHANNEL_OFFSET);
+    if (sound_channels[index - AUDIO_CHANNEL_OFFSET]->mchunk == snd.mc) {
+      Mix_HaltChannel(index - AUDIO_CHANNEL_OFFSET);
+    }
+  }
+  const Sound& snd = sounds.get(index);
+  for(size_t i =0; i <sound_channels.size();i++) {
+    if (sound_channels[i]->mchunk == snd.mc) {
+      Mix_HaltChannel(i);
+    }
+  }
+}
+
+void audio_pause_sound(int index) {
+  if (index >= AUDIO_CHANNEL_OFFSET) {
+    const Sound& snd = sounds.get(index - AUDIO_CHANNEL_OFFSET);
+    if (sound_channels[index - AUDIO_CHANNEL_OFFSET]->mchunk == snd.mc && Mix_Playing(index - AUDIO_CHANNEL_OFFSET) == 1) {
+      Mix_Pause(index - AUDIO_CHANNEL_OFFSET);
+    }
+  }
+  const Sound& snd = sounds.get(index);
+  for(size_t i =0; i <sound_channels.size();i++) {
+    if (sound_channels[i]->mchunk == snd.mc && Mix_Playing(i) == 1) {
+      Mix_Pause(i);
+    }
+  }
+}

```

Comment:
These other functions all require the same fixes as above. You should not be grabbing ANY sound when index is actually a channel, and you should also be returning instead of falling through. This function is actually an if/else in OpenAL system.
https://github.com/enigma-dev/enigma-dev/blob/fad7d5ffb0256d6b330861f5cb39482e14a76823/ENIGMAsystem/SHELL/Audio_Systems/OpenAL/ALadvanced.cpp#L175

Reply:
resolved